### ISSUE TITLE
support joining dependencies together; so that the PR is consistent

### ISSUE THREAD

**Issue by jstrachan** [fabric8io/updatebot#4]
imagine if your project dependency graph looked like this:
![ui-integration-arch2](https://user-images.githubusercontent.com/1184371/30814447-3a444686-a1de-11e7-8caf-9e3cdc40e5ea.png)

when `ngx-base` is released you want to 

* create PRs on the other `ngx-*` repos
* not create a PR on `fabric8-planner` but wait until consistent commits come from all the `ngx-*` libraries and the `ngx-base` commit in a single PR

i.e. releasing `ngx-base` generates 4 releases and 4 dependency changes on `fabric8-planner` but we want that to appear as a single PR when its ready to be tested/merged; not in separate pieces - as those PRs:

* may be invalid due to version mismatch
* create unstable master if merged
* leads to lots of human noise ;)

## Approach

The current approach for _joining_ related dependency changes across project releases is to use github issues.

So that:

* updatebot on `ngx-base` will generate an issue on `fabric8-planner` with the updatebot command to update the `ngx-base` dependencies - as it can detect the version clash between planner's dependencies & knows we need to wait for further dependency update commands from `ngx-widgets, ngx-login-client, ngx-fabric8-wit`
* updatebot on `ngx-widgets & ngx-login-client` will then add the dependency update commands to the updatebot issue on `fabric8-planner`
* updatebot on `ngx-fabric8-wit` (assuming thats the last one to change) - will detect all the other commands on the issue; realise that now things are all consistent and submit a single PR which fixes the issue raised for all the dependency changes

In future we could add commands on the issue to force an early PR despite waiting for future dependency upgrades; plus can use comments to indicate what dependencies the issue is waiting on

**Comment by joshuawilson** [comment]
Currently `ngx-base` is not a part of `ngx-widgets` but I can add it. That would truly make it the "base".

We need the changes to also go to `fabric8-stack-analysis` too, just like `fabric8-planner`. Once both of them are updated then `fabric8-ui` can be updated with everything that has changed.

These sub-components are not directly linked to planner, any component can consume them. `fabric8-ui` consumes them all directly.

**Comment by jstrachan** [comment]
Note that ‘ngx-base’ does not need to be in the other ‘ngx-*’ libraries as a dependency for it to be the exemplar for its own dependencies for, say, angular.

The changes will ripple through all repos we include in the YAML config file (eg use ‘includes: *’ & will do so in dependency tree order only when the versions are consistent

**Comment by joshuawilson** [comment]
ah, very nice

**Comment by jstrachan** [comment]
We've a test case for this which simulates releases on projects (by making git local version changes and providing a stub dependency checker to avoid really doing releases in JUnit ;-):
https://github.com/fabric8io/updatebot/blob/master/updatebot-core/src/test/java/io/fabric8/updatebot/fork/ForkJoinReleaseTest.java#L78-L95

The test case only asserts on the pending issues & PRs on the final git repo, `fabric8-planner`, you'll notice PRs on the other repos get created along the way and we use a github issue to keep track of the pending issues until we can PR them.

```
INFO  Repositories                   - Cloning: jstrachan-testing/ngx-base to /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-base
Cloning into 'ngx-base'...
INFO  Repositories                   - Cloning: jstrachan-testing/ngx-widgets to /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-widgets
Cloning into 'ngx-widgets'...
INFO  Repositories                   - Cloning: jstrachan-testing/ngx-fabric8-wit to /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-fabric8-wit
Cloning into 'ngx-fabric8-wit'...
INFO  Repositories                   - Cloning: jstrachan-testing/ngx-login-client to /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-login-client
Cloning into 'ngx-login-client'...
INFO  Repositories                   - Cloning: jstrachan-testing/fabric8-planner to /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/fabric8-planner
Cloning into 'fabric8-planner'...
INFO  ForkJoinReleaseTest            - Simulating ngx-base release of 2.0.0
[master bac7e42] Simulated release of version 2.0.0
 1 file changed, 1 insertion(+), 1 deletion(-)
INFO  ForkJoinReleaseTest            - Pushing source changes from /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-base
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-base on jstrachan-testing/ngx-widgets has version changes ngx-base => 2.0.0
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-base on jstrachan-testing/ngx-fabric8-wit has version changes ngx-base => 2.0.0
[updatebot-aa54754a-f549-4a5c-8c68-4b1f192beced bdc01f6] fix(versions): git://github.com/jstrachan-testing/ngx-base.git
 1 file changed, 1 insertion(+), 1 deletion(-)
To git@github.com:jstrachan-testing/ngx-fabric8-wit
 * [new branch]      updatebot-aa54754a-f549-4a5c-8c68-4b1f192beced -> updatebot-aa54754a-f549-4a5c-8c68-4b1f192beced
INFO  ModifyFilesCommandSupport      - Created pull request https://github.com/jstrachan-testing/ngx-fabric8-wit/pull/39
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-base on jstrachan-testing/ngx-login-client has version changes ngx-base => 2.0.0
[updatebot-4926ff19-64f9-4782-8ffd-0f79497d1d01 f391547] fix(versions): git://github.com/jstrachan-testing/ngx-base.git
 1 file changed, 1 insertion(+), 1 deletion(-)
To git@github.com:jstrachan-testing/ngx-login-client
 * [new branch]      updatebot-4926ff19-64f9-4782-8ffd-0f79497d1d01 -> updatebot-4926ff19-64f9-4782-8ffd-0f79497d1d01
INFO  ModifyFilesCommandSupport      - Created pull request https://github.com/jstrachan-testing/ngx-login-client/pull/34
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-base on jstrachan-testing/fabric8-planner has version changes ngx-base => 2.0.0
INFO  ModifyFilesCommandSupport      - Created issue https://github.com/jstrachan-testing/fabric8-planner/issues/67
WARN  ForkJoinReleaseTest            - ===> fabric8-planner Issue count: 1 PR count: 0
INFO  Issues                         - Open issue https://github.com/jstrachan-testing/fabric8-planner/issues/67
INFO  ForkJoinReleaseTest            - Simulating ngx-login-client release of 2.2.2
[master 80683b1] Simulated release of version 2.2.2
 1 file changed, 2 insertions(+), 2 deletions(-)
INFO  ForkJoinReleaseTest            - Pushing source changes from /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-login-client
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-login-client on jstrachan-testing/ngx-base has version changes ngx-login-client => 2.2.2
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-login-client on jstrachan-testing/ngx-widgets has version changes ngx-login-client => 2.2.2
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-login-client on jstrachan-testing/ngx-fabric8-wit has version changes ngx-login-client => 2.2.2
[updatebot-83cf2b33-c3fe-44e8-956d-ef6993b2e890 3c49fb2] fix(versions): git@github.com:jstrachan-testing/ngx-login-client
 1 file changed, 1 insertion(+), 1 deletion(-)
To git@github.com:jstrachan-testing/ngx-fabric8-wit
 * [new branch]      updatebot-83cf2b33-c3fe-44e8-956d-ef6993b2e890 -> updatebot-83cf2b33-c3fe-44e8-956d-ef6993b2e890
INFO  ModifyFilesCommandSupport      - Created pull request https://github.com/jstrachan-testing/ngx-fabric8-wit/pull/40
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-login-client on jstrachan-testing/fabric8-planner has version changes ngx-login-client => 2.2.2
INFO  ModifyFilesCommandSupport      - Modifying issue https://github.com/jstrachan-testing/fabric8-planner/issues/67
INFO  ForkJoinReleaseTest            - Simulating ngx-fabric8-wit release of 7.3.3
[master 14b536e] Simulated release of version 7.3.3
 1 file changed, 2 insertions(+), 2 deletions(-)
INFO  ForkJoinReleaseTest            - Pushing source changes from /target/test-data/io/fabric8/updatebot/fork/ForkJoinReleaseTest/github/jstrachan-testing/ngx-fabric8-wit
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-fabric8-wit on jstrachan-testing/ngx-base has version changes ngx-fabric8-wit => 7.3.3
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-fabric8-wit on jstrachan-testing/ngx-widgets has version changes ngx-fabric8-wit => 7.3.3
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-fabric8-wit on jstrachan-testing/ngx-login-client has version changes ngx-fabric8-wit => 7.3.3
INFO  PushSourceChanges              - pushing jstrachan-testing/ngx-fabric8-wit on jstrachan-testing/fabric8-planner has version changes ngx-fabric8-wit => 7.3.3
INFO  ModifyFilesCommandSupport      - Closing issue as we have no further pending issues
[updatebot-a35a1350-7215-4a3b-874e-c747e3e46248 63e2952] fix(versions): git@github.com:jstrachan-testing/ngx-fabric8-wit
 1 file changed, 3 insertions(+), 3 deletions(-)
To git@github.com:jstrachan-testing/fabric8-planner
 * [new branch]      updatebot-a35a1350-7215-4a3b-874e-c747e3e46248 -> updatebot-a35a1350-7215-4a3b-874e-c747e3e46248
INFO  ModifyFilesCommandSupport      - Created pull request https://github.com/jstrachan-testing/fabric8-planner/pull/68
```

**Comment by jstrachan** [comment]
@joshuawilson here's a sample PR generated by the test case: https://github.com/jstrachan-testing/fabric8-planner/pull/92 so it waits until 3 releases have happened before submitting a single PR against `fabric8-planner`. Until that time its using this issue to track the dependency conflicts:
https://github.com/jstrachan-testing/fabric8-planner/issues/91

both the issue & PR are linked with comments on both to try describe whats happening and why.

In the future we could get more smart, like adding comment commands to the issue to ignore some dependency conflicts; or spin off some of them as a separate PR or something
